Status bar count

UziTech opened the pull request:
### Description of the Change

Add a notifications count to the status bar that will open the notifications log when clicked.

![status bar count](https://user-images.githubusercontent.com/97994/28998231-813edfbc-79eb-11e7-8bc9-8d97153f4243.gif)

The color of the status bar element indicates what type of notification was last received.

The tooltip on the status bar element will show the last time a notification was recieved

![image](https://user-images.githubusercontent.com/97994/33899136-30afaf54-df30-11e7-8b32-46dd644b087a.png)


### Benefits

Have an easy way to toggle the notifications log.

See the number of notifications.

See when the last notification was received and what type of notification it is.

### Possible Drawbacks

some people might not want it taking up space. (They can easily hide it in styles.less)

### Applicable Issues

https://github.com/atom/notifications/issues/29#issuecomment-110062482

50Wliu commented on lib/status-bar-manager.coffee:
```diff
@@ -0,0 +1,61 @@
+{CompositeDisposable, Disposable} = require 'atom'
```
Would you mind converting this and the new spec file to JavaScript?  We're trying to avoid adding any new Coffeescript files.

50Wliu commented on lib/main.coffee:
```diff
@@ -113,13 +117,18 @@ Notifications =
     notification.setDisplayed(true)
     @lastNotification = notification
 
+  statusBarService: (statusBar) ->
```
The convention followed by other Atom packages is to name this `consumeStatusBar`.

50Wliu commented on lib/main.coffee:
```diff
@@ -113,13 +117,18 @@ Notifications =
     @lastNotification = notification
 
+  statusBarService: (statusBar) ->
+    @statusBarManager = new StatusBarManager(statusBar, @duplicateTimeDelay)
```
Similarly, the convention would for this to be `NotificationStatusView`.

50Wliu commented on spec/notifications-status-view-spec.js:
```diff
@@ -0,0 +1,100 @@
+  describe("when there are notifications before activation", () => {
+    beforeEach(() =>
+      waitsForPromise(() =>
+        // Wrapped in Promise.resolve so this test continues to work on earlier versions of Atom
```
Good news: this isn't needed anymore :)

And if you want you can try converting this spec file to use async/await (see https://github.com/atom/archive-view/pull/57 for example), but it's not necessary.

I'll try to give an actual content review in the next few days.

UziTech replied:
It looks like it only works with the [async-spec-helper.js](https://github.com/atom/archive-view/pull/57/files#diff-41ed07bdd040722c36a531c9e36219ff) file in the pr

did you want me to add that file?

It would be really nice if we could move to jasmine 2 with [atom-jasmine2-test-runner](https://www.npmjs.com/package/atom-jasmine2-test-runner)

That has built in support for async/await and more

UziTech replied:
atom-jasmine2-test-runner even works with jasmine 1 specs so we wouldn't have to move them all over at once

50Wliu commented on lib/main.coffee:
```diff
@@ -113,13 +117,18 @@ Notifications =
     @lastNotification = notification
 
+  consumeStatusBar: (statusBar) ->
+    @statusBarManager = new NotificationsStatusView(statusBar, @duplicateTimeDelay)
```
`notificationsStatusView`

UziTech replied:
Shouldn't that be uppercase since it is a class like `NotificationElement` and `NotificationsLog`

50Wliu replied:
oh sorry, I meant changing `statusBarManager` to `notificationsStatusView`.

UziTech replied:
that makes more sense

50Wliu commented on lib/notifications-status-view.js:
```diff
@@ -0,0 +1,76 @@
+    constructor(statusBar, duplicateTimeDelay) {
+      this.count = 0;
+      this.duplicateTimeDelay = duplicateTimeDelay;
+      this.subscriptions = new CompositeDisposable;
```
`new CompositeDisposable()`

50Wliu commented on lib/notifications-status-view.js:
```diff
@@ -0,0 +1,76 @@
+    render() {
+      this.number = document.createElement('div');
+      this.number.textContent = this.count;
+      this.number.addEventListener('animationend', (e) => {
```
I'm worried that this will be too distracting.  Did you think the number changing was too subtle to notice without an animation?

UziTech replied:
Ya, sometimes I like turning the notification popups off when I am trying to concentrate and that seemed like a good way not give it a little pop.

Also that way people can easily turn it off or specify a different animation by adding the following to their `styles.less`

```css
.notifications-count .new-notification {
  animation: none;
}
```

UziTech replied:
this could allow other packages like [`notifications-plus-confetti`](https://atom.io/packages/notifications-plus-confetti)

![notifications-plus-confetti](https://i.github-camo.com/fd3e96e1d555406b26812ccaa3a10cfee72f5f27/68747470733a2f2f757365722d696d616765732e67697468756275736572636f6e74656e742e636f6d2f39373939342f32393233383431372d62356134333332342d376566612d313165372d383362362d3065636636343666363736362e676966)

50Wliu commented on lib/notifications-status-view.js:
```diff
@@ -0,0 +1,76 @@
+      this.element = document.createElement('a');
+      this.element.classList.add('notifications-count', 'inline-block');
+      this.tooltip = atom.tooltips.add(this.element, {title: 'Notifications'});
+      let span = document.createElement('span');
```
const

50Wliu commented on lib/notifications-status-view.js:
```diff
@@ -0,0 +1,76 @@
+      let span = document.createElement('span');
+      span.appendChild(this.number);
+      this.element.appendChild(span);
+      this.element.addEventListener('click', () => {
```
this should have an associated removeEventListener

50Wliu commented on lib/notifications-status-view.js:
```diff
@@ -0,0 +1,76 @@
+        atom.commands.dispatch(this.element, 'notifications:toggle-log')
+      });
+
+      let lastNotification = null;
```
Isn't this already handled in https://github.com/UziTech/notifications/blob/5191c3264ea7e2e43c4b2cb3aec45d650111a8c9/lib/main.coffee#L114-L120?

I noticed I missed this similar piece of code in my review of the notifications log, sorry for not mentioning it earlier.

UziTech replied:
yes this handles adding notifications that have happened before initializing the statusbar.

Without this code here, if the user disables and re-enables the statusbar package the count would be set to 0 even if there are notifications in the log.

UziTech replied:
I guess I could just pass the notificationsLog item count and the last notification to get the type and message for the tooltip

50Wliu commented on lib/notifications-status-view.js:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    addNotification(notification) {
+      let date = moment(notification.timestamp).format('LT');
```
const

50Wliu commented on lib/notifications-status-view.js:
```diff
@@ -0,0 +1,76 @@
+    addNotification(notification) {
+      let date = moment(notification.timestamp).format('LT');
+      this.tooltip.dispose();
+      this.tooltip = atom.tooltips.add(this.element, {title: `Last Notification ${date}`});
```
I'm not sure how useful this tooltip would be.  Alternatives could include:
* Displaying the number of each type of notification
* Displaying the title of the last notification

UziTech replied:
I wasn't sure what to put here but i thought it should be something useful.

I like the idea of the last notification title

50Wliu commented on lib/notifications-status-view.js:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    destroy() {
+      this.subscriptions.dispose();
```
It looks like you only added one disposable, that being the removal of the tile.  What do you think about explicitly stating `this.tile.destroy()` here instead?

UziTech replied:
This is more or less just to future proof the code in the event that there are more subscriptions (like the comment on line 17 about `onDidClearNotifications`)

50Wliu commented:
I think that the status bar tile should not show up until the initial notification is created.

50Wliu commented on lib/notifications-status-view.js:
```diff
@@ -0,0 +1,76 @@
+      this.number.textContent = this.count;
+      this.element.removeAttribute('last-type');
+      this.tooltip.dispose();
+      this.tooltip = atom.tooltips.add(this.element, {title: "Notifications"});
```
Could read "No notifications" instead.

50Wliu commented:
Oh, also remove the package-lock.json please.

UziTech commented:
You could add

```css
.notifications-count:not([last-type]) {
  display: none;
}
```

to `styles.less` to hide it when the notification count is 0

UziTech commented:
I think I addressed all of your comments. Let me know if I missed something or if you want me to change anything else.

50Wliu commented on lib/notifications-status-view.js:
```diff
@@ -0,0 +1,87 @@
+
+      this.element = document.createElement('a');
+      this.element.classList.add('notifications-count', 'inline-block');
+      this.tooltip = atom.tooltips.add(this.element, {title: 'Notifications'});
```
This tooltip should be consistent with the other tooltip states.  That is, if there are existing notifications when the tile is instantiated it should show the last message, otherwise it should show "No notifications".

50Wliu commented:
I tried this out and I still think the notification animation is too distracting.  I think by default it should have no animation as no other status bar tiles have animations (with the exception of "busy"-style tiles).

/cc @simurai for some UI/UX inspiration 🐣

50Wliu commented:
Also, I realized that it's possible that people could interpret the following as a bug with the tooltip and not that it's the notification message:
![tooltip-uncaught-error-message](https://user-images.githubusercontent.com/2766036/35774530-392fbe46-0940-11e8-9683-2450c2d550e4.png)

UziTech commented:
I removed the animation ☹️ but I guess someone could just use `atom.notifications.onDidDisplay()` to add confetti or whatever.

I also added the same .eslintrc as atom/tabs since we are starting to move to .js 😃

simurai commented:
> @UziTech I removed the animation ☹️ but I guess someone could just use `atom.notifications.onDidDisplay()` to add confetti or whatever.

Theme authors could use the `[last-type='...']` selectors to trigger custom animations. For example only do some wiggling when the type changes to `fatal`. -> `[last-type='fatal'] { ... }`.

---

> @50Wliu I think that the status bar tile should not show up until the initial notification is created.

I agree with only showing the count if above `0`. Don't think people would want to open the log if the list is empty. And it looks more cleaned up when opening a new Atom window.

---

There is a mockup from a very long time ago: https://github.com/atom/atom/issues/4212#issuecomment-63917665. Maybe having an icon + count could be nice and good for accessibility? I'll see if I can play around with it in a separate PR.

simurai commented:
Here an [alternative](https://github.com/UziTech/notifications/pull/3) with a filled background, instead of the border:

![screen shot 2018-02-05 at 4 02 45 pm](https://user-images.githubusercontent.com/378023/35792264-99aae3a4-0a8f-11e8-9f44-f4bf44a688e3.png)

ps. I just realized that I moved it in front of the updates. Only in the DevTools to see how that feels.

UziTech commented:
I like it. Looks beautiful @simurai 💯 

Do we want to add css to hide it when there are no notifications?

I personally think it would be better to show "0". That way if someone is looking for how many notifications there are they will see "0" instead of looking around for it for a bit to make sure they didn't miss it. Seems like less cognitive load to always have the information directly available.

UziTech commented:
probably the same reason the github package leaves the status bar element when there are 0 files changed

UziTech commented:
Also since the status-bar icon acts as a switch to toggle the log, we would want to keep it visible so when they [clear](https://github.com/atom/notifications/pull/180) the log they have a common button to hide the panel.

50Wliu commented on lib/notifications-status-view.js:
```diff
@@ -0,0 +1,79 @@
+
+    this.element = document.createElement('a')
+    this.element.classList.add('notifications-count', 'inline-block')
+    const s = (this.count === 1 ? '' : 's')
```
:+1: Thanks for not pulling in underscore-plus for this

50Wliu commented:
Okay, I think this is almost ready to go!

* Most Atom repositories that use JS are using Standard, not ESLint (tabs seems to be an exception).  See for example the Standard config in [update-package-dependencies](https://github.com/atom/update-package-dependencies/blob/a2d4ca555a71f1d4314ab01e8e41db2108e2f203/package.json#L28).
* I know that since you wrote and maintain the Jasmine2 test runner for Atom you'd like to use it here, but I would appreciate it if you could just stick to the legacy spec runner for now and copy over `async-spec-helpers.js` from [atom/atom](https://github.com/atom/atom/blob/d5bd4421715a522a03e96725edba125dbb304f02/spec/async-spec-helpers.js) for consistency with other repos.

UziTech commented:
I agree with using the legacy spec runner, but haven't we learned our lesson with using abstractions from coffeescript. Standard still doesn't support some useful features of eslint. And any time eslint is updated you will have to wait for standard to update as well. I'm all for using whatever eslint configuration you want but could we use the actual eslint to prevent another unmaintained abstraction from leaking in?

simurai commented:
> @UziTech Seems like less cognitive load to always have the information directly available.

Well, it could also be **more** cognitive load always having to look at the `0` when trying to find something else. :stuck_out_tongue_winking_eye: And I think it's a bit different from things like a chat notification, that you want to check from time to time. I see the Log more as "a place to go back, in case I missed something" and less "a place I want to keep an eye on".

> Also since the status-bar icon acts as a switch to toggle the log, we would want to keep it visible so when they clear the log they have a common button to hide the panel.

👍  Ok, maybe that's the best argument too keep it. :grinning:

Also, if [Persistent Controls for Docks](https://github.com/atom/atom/issues/16116) will ever make it, then this question will be answered for good.

UziTech commented:
Persistent Controls for Docks looks awesome. That would free up about half of my status bar 😄

50Wliu commented on spec/notifications-status-view-spec.js:
```diff
@@ -0,0 +1,84 @@
+/** @babel */
```
Don't use Babel.  Pretty sure this just entails changing the imports to requires.

50Wliu commented:
I don't want to hold this PR up over something like linting so we'll ship it with ESLint and deal with that later.  One more comment about removing Babel from the main spec file.

I've been using this for a few days and have another UI suggestion.  If an error pops up, then the status bar icon stays red until another notification pops up; I can see that getting distracting.  Clearly this will get better when clear support is added, but I think it's something that we still might want to look into. 
 Possible solutions:
* Always use neutral-gray
* Reset the counter to 0 when the log is opened

Something else that I'm missing?

UziTech commented:
@50Wliu do you want me to remove `\** @babel *\` from `async-spec-helpers.js` as well?

> Clearly this will get better when clear support is added

I believe this should be left until the logs are cleared.

UziTech commented:
It seems like there should be a better way to write tests for core packages than duplicating a file. Atom could create an `atom-async-spec-helper` npm package like [`jasmine-fix`](https://www.npmjs.com/package/jasmine-fix)

simurai commented:
> @50Wliu If an error pops up, then the status bar icon stays red until another notification pops up; I can see that getting distracting.

Also, showing a counter like:

![image](https://user-images.githubusercontent.com/378023/36404027-e62173cc-162b-11e8-9ac5-ae6bb2c839f6.png)

might suggest that there are 13 errors, and not that just the _last_ notification is an error.

> Reset the counter to 0 when the log is opened

👍 Yeah, I like that one. Then it acts more like a "unread count".

UziTech commented:
If it is going to be like an unread counter, I don't think it would matter if the user thinks there are 13 errors or just the last one is an error.

I like the unread counter idea. More like other notification systems. Then there should be a way to mark the notifications as read without clearing them.

UziTech commented:
we could do something like atom ide diagnostics
![image](https://user-images.githubusercontent.com/97994/36405123-87ab073a-15b4-11e8-8390-66b280913f28.png)

UziTech commented:
@simurai @50Wliu Do we want to go with the ide diagnostics design?

jasonrudolph commented:
:wave: Hi @UziTech! Thanks for your perseverance on this pull request. I see that it's had quite a bit of back and forth over the past year and a half. I'm sorry we haven't been able to shepherd this through the review process faster. 

I want to get this unblocked for you. How would you feel about shipping this feature as a package? For example, it might be called something like `notifications-status-badge` or `notifications-status-bar`. I see at least a few benefits to that approach:

- You could ship this feature without having to wait for review from the Atom team :shipit:
- The core `notifications` package wouldn't gain an additional feature for the Atom team to support/maintain :innocent:
- The core `notifications` package wouldn't introduce the potential drawback stated in the pull request body: "Some people might not want [the counter] taking up space. (They can easily hide it in styles.less)." 

What do you think?

UziTech commented:
That works thanks for the suggestion
